core/Segment.php

```diff
@@ -460,8 +465,42 @@ public function getHash()
 
     public static function getSegmentHash($definition)
     {
-        // urldecode to normalize the string, as browsers may send slightly different payloads for the same archive
-        return md5(urldecode($definition));
+        $cache = Cache::getEagerCache();
+        $cacheKey = self::CACHE_KEY . md5($definition);
+
+        if ($cache->contains($cacheKey)) {
+            return $cache->fetch($cacheKey);
+        }
+
+        $defaultHash = md5(urldecode($definition));
+
+        // if the cache for segments already built, but this segment was not found,
+        // we return the default segment, this can be a segment from url or
+        // something like "visitorType==new"
+        if ($cache->contains(self::SEGMENT_HAS_BUILT_CACHE_KEY)) {
+            return $defaultHash;
+        }
+
+        // the segment hash is not built yet, let's do it
+        $cache->save(self::SEGMENT_HAS_BUILT_CACHE_KEY, true);
+
+        $model = new SegmentEditorModel();
+        $segments = $model->getAllSegmentsForAllUsers();
```

Comment:
@flamisz to remove the amount of data and speed things up a bit we could add a new query to only fetch `hash` and `definition`. It's not too super important though unless there were really heaps of entries. So maybe for now we skip this as it's likely not needed.